pkg/pilot/genericpilot/genericpilot.go

```diff
@@ -83,28 +43,44 @@ func (g *GenericPilot) Run() error {
 	// setup healthz handlers
 	g.serveHealthz()
 
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		wait.Until(g.worker, time.Second, g.Options.StopCh)
-	}()
+	ctrlStopCh := make(chan struct{})
+	defer close(ctrlStopCh)
+	go g.controller.Run(ctrlStopCh)
+
+	// block until told to shutdown
+	select {
+	case <-g.Options.StopCh:
+	case <-g.processExitChan():
+		if !g.process.State().Success() {
+			glog.V(4).Infof("Underlying process failed")
+		} else {
+			glog.V(4).Infof("Underlying process unexpectedly exited")
+		}
+	}
 
-	<-g.Options.StopCh
 	glog.V(4).Infof("Shutdown signal received")
-	// set g.shutdown = true to signal preStop hooks to run
-	g.shutdown = true
-	glog.V(4).Infof("Waiting for process exit and hooks to execute")
-	// wait until postStop hooks have run
-	wait.Poll(time.Second*1, time.Minute*10, func() (bool, error) {
-		return g.lastCompletedPhase == v1alpha1.PilotPhasePostStop, nil
-	})
-	glog.V(4).Infof("Shutting down workqueue")
-	// shutdown the worker queue
-	g.queue.ShutDown()
-	glog.V(4).Infof("Shutting down generic pilot controller workers")
-	// wait for workers to exit
-	wg.Wait()
-	glog.V(4).Infof("Generic pilot controller workers stopped")
-	return nil
+	thisPilot, err := g.controller.ThisPilot()
+	if err != nil {
+		return err
+	}
+	return g.stop(thisPilot)
+}
+
+func (g *GenericPilot) processExitChan() <-chan struct{} {
+	out := make(chan struct{})
+	go func() {
+		defer close(out)
+		// don't call wait until the process is actually created
+		for {
+			// wait until the proces is running before calling Wait
+			if g.process != nil && g.process.Running() {
+				break
+			}
+			time.Sleep(2)
+		}
+		// we must call Wait else process.ProcessState won't be populated with
+		// exit code/details of the process.
+		g.process.Wait()
```

Comment:
This also returns an error, which needs handling.

Reply:
This is now resolved as part of the later commit which improves processmanager